evm/eth.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		ctx.Json(http.StatusBadRequest, &CallResponse{Err: jsonrpc.Err(jsonrpc.InvalidJSON, err.Error())})
+		return
+	}
+
+	cfg := s.cfg
+	setDefaults(cfg)
+	address := callReq.Address
+	input := callReq.Input
+
+	var (
+		vmenv   = NewEnv(cfg)
+		sender  = vm.AccountRef(cfg.Origin)
+		statedb = cfg.State
+		rules   = cfg.ChainConfig.Rules(vmenv.Context.BlockNumber, vmenv.Context.Random != nil, vmenv.Context.Time)
+	)
+	if cfg.EVMConfig.Tracer != nil && cfg.EVMConfig.Tracer.OnTxStart != nil {
+		cfg.EVMConfig.Tracer.OnTxStart(vmenv.GetVMContext(), types.NewTx(&types.LegacyTx{To: &address, Data: input, Value: cfg.Value, Gas: cfg.GasLimit}), cfg.Origin)
+	}
+	// Execute the preparatory steps for state transition which includes:
+	// - prepare accessList(post-berlin)
+	// - reset transient storage(eip 1153)
+	statedb.Prepare(rules, cfg.Origin, cfg.Coinbase, &address, vm.ActivePrecompiles(rules), nil)
+
+	// Call the code with the given configuration.
+	ret, leftOverGas, err := vmenv.Call(
+		sender,
+		address,
+		input,
+		cfg.GasLimit,
+		uint256.MustFromBig(cfg.Value),
+	)
+	println("Return ret value:", ret)
+	println("Return leftOverGas value:", leftOverGas)
+
+	if err != nil {
+		ctx.Json(http.StatusInternalServerError, &CallResponse{Err: jsonrpc.Err(jsonrpc.InternalError, err.Error())})
+		return
+	}
+
+	ctx.JsonOk(&CallResponse{Ret: ret, LeftOverGas: leftOverGas})
+}
+
+// Create executes the code using the EVM create method
+func (s *Solidity) Create(ctx *context.WriteContext) error {
+	txCreate := new(CreateRequest)
+	err := ctx.BindJson(txCreate)
+	if err != nil {
+		return err
+	}
+
+	cfg := s.cfg
+	setDefaults(cfg)
+
+	input := txCreate.Input
+
+	if cfg.State == nil {
+		cfg.State, _ = state.New(types.EmptyRootHash, state.NewDatabase(rawdb.NewMemoryDatabase()), nil)
+	}
+	var (
+		vmenv  = NewEnv(cfg)
+		sender = vm.AccountRef(cfg.Origin)
+		rules  = cfg.ChainConfig.Rules(vmenv.Context.BlockNumber, vmenv.Context.Random != nil, vmenv.Context.Time)
+	)
+	if cfg.EVMConfig.Tracer != nil && cfg.EVMConfig.Tracer.OnTxStart != nil {
+		cfg.EVMConfig.Tracer.OnTxStart(vmenv.GetVMContext(), types.NewTx(&types.LegacyTx{Data: input, Value: cfg.Value, Gas: cfg.GasLimit}), cfg.Origin)
+	}
+	// Execute the preparatory steps for state transition which includes:
+	// - prepare accessList(post-berlin)
+	// - reset transient storage(eip 1153)
+	cfg.State.Prepare(rules, cfg.Origin, cfg.Coinbase, nil, vm.ActivePrecompiles(rules), nil)
+	// Call the code with the given configuration.
+	code, address, leftOverGas, err := vmenv.Create(
+		sender,
+		input,
+		cfg.GasLimit,
+		uint256.MustFromBig(cfg.Value),
+	)
+
+	println("Return code value:", code)
+	println("Return address value:", address.Bytes())
+	println("Return leftOverGas value:", leftOverGas)
+
+	return err
+}
+
+func (s *Solidity) Commit(block *yu_types.Block) {
+	blockNumber := uint64(block.Height)
+	statedb, err := s.ethState.NewStateDB(common.Hash(block.StateRoot))
```

Comment:
Don't `NewStateDB()` here, here is only `commit()`, should `NewStateDB()` in other place.  I will push some codes to fix it.